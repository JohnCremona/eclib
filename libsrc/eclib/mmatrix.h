// mmatrix.h: declarations of multiprecision integer matrix class
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
#if     !defined(_ECLIB_MMATRIX_H)
#define _ECLIB_MMATRIX_H      1       //flags that this file has been included

#include "mvector.h"
#include "matrix.h"

#undef scalar
#undef vec
#undef mat
#undef subspace
#undef svec
#undef smat
#undef smat_elim

#define scalar bigint
#define vec vec_m
#define mat mat_m
#define subspace subspace_m
#define svec svec_m
#define smat smat_m
#define smat_elim smat_m_elim

#include "mat.h"

#undef scalar
#undef vec
#undef mat
#undef subspace
#undef svec
#undef smat
#undef smat_elim

mat_m to_mat_m(const mat_i& v);
mat_m to_mat_m(const mat_l& v);
mat_i to_mat_i(const mat_m& v);
mat_l to_mat_l(const mat_m& v);

#if(0)

int liftmat(const mat_m& mm, const bigint& pr, mat_m& m, bigint& dd, int trace=0);
int lift(const msubspace& s, const bigint& pr, msubspace& ans, int trace=0);

class mat_m {
  friend class msubspace;
public:
  // constructors
  mat_m(long nr=0, long nc=0)   :nro(nr), nco(nc) {entries.resize(nr*nc, bigint(0));}
  mat_m(long nr, long nc, const vector<bigint> ent) : nro(nr), nco(nc), entries(ent) {}
  mat_m(const mat_m& m) :nro(m.nro), nco(m.nco){ entries = m.entries;}  // copy constructor
  explicit mat_m(const mat_i&);
  explicit mat_m(const mat_l&);

  // member functions & operators
  void init(long nr=0, long nc=0);
  mat_m& operator=(const mat_m&);            // assignment with copy
  bigint& operator()(long i, long j);        // returns ref to (i,j) entry
  bigint operator()(long i, long j) const;   // returns (i,j) entry
  mat_m slice(long,long,long=-1,long=-1) const;// returns submatrix
  bigint sub(long i, long j) const;             // returns the (i,j) entry
  vec_m row(long i) const;                // returns row i (as a vector)
  vec_m col(long j) const;                // returns col j (as a vector)
  void set(long i, long j, const bigint& x);     // sets the (i,j) entry to x
  void add(long i, long j, const bigint& x);  // adds x to the (i,j) entry
  void setrow(long i, const vec_m& v);
  void setcol(long i, const vec_m& v);
  void swaprows(long r1, long r2);
  void multrow(long r, const bigint& scal);
  void divrow(long r, const bigint& scal);
  bigint content() const;
  bigint row_content(long r) const;
  void clearrow(long r);
  void makeprimitive();
  void operator+=(const mat_m&);
  void operator-=(const mat_m&);
  void operator*=(const bigint&);
  void operator/=(const bigint&);
  // shortens to a matrix o ints or longs if possible
  //the parameter here is a dummy just to distinguish these
  mat_i  shorten(int) const;
  mat_l shorten(long) const;
  long nrows() const {return nro;}
  long ncols() const {return nco;}
  long rank() const;
  long nullity() const;
  bigint trace() const;
  vector<bigint> charpoly() const;
  bigint determinant() const;

  // non-member (friend) functions and operators
  friend mat_m operator*(const mat_m&, const mat_m&);
  friend vec_m operator*(const mat_m&, const vec_m&);
  friend int operator==(const mat_m&, const mat_m&);
  friend ostream& operator<< (ostream&s, const mat_m&);
  friend istream& operator>> (istream&s, mat_m&);
  friend mat_m colcat(const mat_m& a, const mat_m& b);
  friend mat_m rowcat(const mat_m& a, const mat_m& b);
  friend mat_m directsum(const mat_m& a, const mat_m& b);
  friend void elimrows(mat_m& m, long r1, long r2, long pos); //plain elimination, no clearing
  friend void elimrows1(mat_m& m, long r1, long r2, long pos); //elimination + clearing
  friend void elimrows2(mat_m& m, long r1, long r2, long pos, const bigint& last); //elimination + divide by last pivot
  friend mat_m echelon0(const mat_m& m, vec_i& pcols, vec_i& npcols,
                        long& rk, long& ny, bigint& d);
  friend void elimp(mat_m& m, long r1, long r2, long pos, const bigint& pr);
  friend mat_m echelonp(const mat_m& m1, vec_i& pcols, vec_i& npcols,
                        long& rk, long& ny, bigint& d, const bigint& pr);
  friend mat_m echmodp(const mat_m& m, vec_i& pcols, vec_i& npcols,
                       long& rk, long& ny, const bigint& pr);
  friend msubspace combine(const msubspace& s1, const msubspace& s2);
  friend mat_m restrict_mat(const mat_m& m, const msubspace& s);
  friend int liftmat(const mat_m& mm, const bigint& pr, mat_m& m, bigint& dd, int trace);
  friend int lift(const msubspace& s, const bigint& pr, msubspace& ans, int trace);
  friend msubspace pcombine(const msubspace& s1, const msubspace& s2, const bigint& pr);
  friend mat_m prestrict(const mat_m& m, const msubspace& s, const bigint& pr);
  friend mat_m matmulmodp(const mat_m&, const mat_m&, const bigint& pr);

  // Implementation
private:
  long nro,nco;
  vector<bigint> entries;  // stored in one array of size nro+nco, by rows
};

// Declaration of non-friend functions

mat_m operator+(const mat_m&);                   // unary
mat_m operator-(const mat_m&);                   // unary
mat_m operator+(const mat_m& m1, const mat_m& m2);
mat_m operator-(const mat_m& m1, const mat_m& m2);
mat_m operator*(const bigint& scal, const mat_m& m);
mat_m operator/(const mat_m& m, const bigint& scal);
int operator!=(const mat_m& m1, const mat_m& m2);
mat_m midmat(long n);  // = multi-idmat
mat_m transpose(const mat_m& m);
// submatrix of rows indexed by v, all columns
mat_m rowsubmat(const mat_m& m, const vec_i& v);
// submatrix of rows indexed by iv, columns indexed by jv
mat_m submatrix(const mat_m& m, const vec_i& iv, const vec_i& jv);
mat_m echelon(const mat_m& m, vec_i& pcols, vec_i& npcols,
                          long& rk, long& ny, bigint& d, int method=0);
mat_m echelon(const mat_m& m, vec_l& pcols, vec_l& npcols,
                          long& rk, long& ny, bigint& d, int method=0);
mat_m addscalar(const mat_m&, const bigint&);
vec_m apply(const mat_m&, const vec_m&);

#endif // end of #if(0)

#endif
