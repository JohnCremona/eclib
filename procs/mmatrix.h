// mmatrix.h: declarations of multiprecision integer matrix class
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2005 John Cremona
// 
// This file is part of the mwrank package.
// 
// mwrank is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// mwrank is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with mwrank; if not, write to the Free Software Foundation,
// Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
// 
//////////////////////////////////////////////////////////////////////////
 
#if     !defined(_MMATRIX_H)
#define _MMATRIX_H      1       //flags that this file has been included

#include "mvector.h"
#include "matrix.h"

class mat_m {
friend class msubspace;
public:
     // constructors
        mat_m(long nr=0, long nc=0);
        mat_m(const mat_m&);                    // copy constructor
        mat_m(const mat_i&);
        mat_m(const mat_l&);
     // destructor
        ~mat_m();

     // member functions & operators
        void init(long nr=0, long nc=0);
        mat_m& operator=(const mat_m&);          // assignment with copy
        bigint& operator()(long i, long j) const;   // returns ref to (i,j) entry
        mat_m slice(long,long,long=-1,long=-1) const;// returns submatrix
        bigint sub(long i, long j) const;             // returns the (i,j) entry
        vec_m row(long i) const;                // returns row i (as a vector)
        vec_m col(long j) const;                // returns col j (as a vector)
        void set(long i, long j, const bigint& x);     // sets the (i,j) entry to x
        void add(long i, long j, const bigint& x);  // adds x to the (i,j) entry  
        void setrow(long i, const vec_m& v);
        void setcol(long i, const vec_m& v);
        void swaprows(long r1, long r2);
        void multrow(long r, const bigint& scal);
        void divrow(long r, const bigint& scal);
        void clearrow(long r);
        mat_m& operator+=(const mat_m&);
        mat_m& operator-=(const mat_m&);
        mat_m& operator*=(const bigint&);
        mat_m& operator/=(const bigint&);
  // shortens to a matrix o ints or longs if possible
  //the parameter here is a dummy just to distinguish these
        mat_i  shorten(int) const;
        mat_l shorten(long) const;

     // non-member (friend) functions and operators
        friend long nrows(const mat_m&);      
        friend long ncols(const mat_m&);      
        friend mat_m operator*(const mat_m&, const mat_m&);
        friend vec_m operator*(const mat_m&, const vec_m&);
        friend int operator==(const mat_m&, const mat_m&);
        friend ostream& operator<< (ostream&s, const mat_m&);
        friend istream& operator>> (istream&s, mat_m&);
        friend mat_m colcat(const mat_m& a, const mat_m& b);
        friend mat_m rowcat(const mat_m& a, const mat_m& b);
        friend mat_m directsum(const mat_m& a, const mat_m& b);
        friend void elimrows(mat_m& m, long r1, long r2, long pos); //plain elimination, no clearing
        friend void elimrows1(mat_m& m, long r1, long r2, long pos); //elimination + clearing
        friend void elimrows2(mat_m& m, long r1, long r2, long pos, const bigint& last); //elimination + divide by last pivot
        friend mat_m echelon0(const mat_m& m, vec_i& pcols, vec_i& npcols,
                                  long& rk, long& ny, bigint& d);
        friend void elimp(const mat_m& m, long r1, long r2, long pos, const bigint& pr);
        friend mat_m echmodp(const mat_m& m, vec_i& pcols, vec_i& npcols,
                                  long& rk, long& ny, const bigint& pr);
        friend msubspace combine(const msubspace& s1, const msubspace& s2);
        friend mat_m restrict(const mat_m& m, const msubspace& s);
        friend msubspace lift(const msubspace& s, const bigint& pr, int =0);
        friend msubspace pcombine(const msubspace& s1, const msubspace& s2, const bigint& pr);
        friend mat_m prestrict(const mat_m& m, const msubspace& s, const bigint& pr);
        friend mat_m matmulmodp(const mat_m&, const mat_m&, const bigint& pr);

// Implementation
private:
       long nro,nco;
       bigint * entries;  // stored in one array, by rows
};



// Declaration of non-friend functions

mat_m operator+(const mat_m&);                   // unary
mat_m operator-(const mat_m&);                   // unary
mat_m operator+(const mat_m& m1, const mat_m& m2);
mat_m operator-(const mat_m& m1, const mat_m& m2);
mat_m operator*(const bigint& scal, const mat_m& m);
mat_m operator/(const mat_m& m, const bigint& scal);
int operator!=(const mat_m& m1, const mat_m& m2);
mat_m midmat(long n);  // = multi-idmat
mat_m transpose(const mat_m& m);
mat_m submatrix(const mat_m& m, const vec_i& iv, const vec_i& jv);
mat_m echelon(const mat_m& m, vec_i& pcols, vec_i& npcols,
                          long& rk, long& ny, bigint& d, int method=0);
mat_m echelon(const mat_m& m, vec_l& pcols, vec_l& npcols,
                          long& rk, long& ny, bigint& d, int method=0);
long rank(const mat_m&);
long nullity(const mat_m&);
bigint trace(const mat_m&);
vector<bigint> charpoly(const mat_m&);
bigint determinant(const mat_m&);
mat_m addscalar(const mat_m&, const bigint&);
vec_m apply(const mat_m&, const vec_m&);

#endif
