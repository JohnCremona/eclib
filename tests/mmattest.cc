// mmattest.cc: Multiprecision matrix package test program
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2005 John Cremona
// 
// This file is part of the mwrank package.
// 
// mwrank is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// mwrank is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with mwrank; if not, write to the Free Software Foundation,
// Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
// 
//////////////////////////////////////////////////////////////////////////
 
#include <time.h>
#include "mmatrix.h"

int main(void)
{
  time_t starttime,stoptime;
  time(&starttime);
  cout << "\nMultiprecision matrix package test program.\n\n";
  {
    long i,j; 
    scalar r;
    mat_m a,aug,ref;
    vec_l pc(1),npc(1); vec_m poly(1);
    {
      cout << "Enter size of a square matrix A: "; cin >> r;
      a.init(r,r);
      cout << "Enter entries of A: "; cin >> a;
      cout << "A = " << a;
    }

//  The following should work, doesn't in TURBO C++, but DOES with GCC!!

{
cout << "Creating an array of 3 matrices\n";
mat_m* matlist = new mat_m[3];
matlist[0] = a;
matlist[1] = BIGINT(2)*a;
matlist[2] = BIGINT(3)*a;
cout << " A=" << matlist[0];
cout << "2A=" << matlist[1];
cout << "3A=" << matlist[2];
delete[] matlist;
}

{
for (i=1; i<=r; i++)
 cout << "row(A,"<<i<<") = " << a.row(i) << endl;
cout << "A = " << a;
for (j=1; j<=r; j++)
 cout << "col(A,"<<j<<") = " << a.col(j) << endl;
cout << "A = " << a;
cout << "directsum(A,A) = " << directsum(a,a);
cout << "Enter any number "; cin >> i;
}
{
mat sa = a.shorten(r);
cout << "After shortening to a matrix of longs, A = " << sa;
}
mat_m b = a;
{
cout << "B = A = " << b;
cout << "Enter any number "; cin >> i;
cout << "B==A?" << (b==a) << endl;
cout << "B!=A?" << (b!=a) << endl;
b+=a;
cout << "after B+:=A, A = " << a << "and B = " << b;
cout << "Enter any number "; cin >> i;
}
{
b-=a;
cout << "after B-:=A, A = " << a << "and B = " << b;
cout << "Enter any number "; cin >> i;
bigint two; two=2;
b*=two;
cout << "after B*:=2, A = " << a << "and B = " << b;
cout << "Enter any number "; cin >> i;
b/=two;
cout << "after B/:=2, A = " << a << "and B = " << b;
cout << "Enter any number "; cin >> i;
}
{
cout << "A+B=" << (a+b);
cout << "Now A = " << a << "and B = " << b;
cout << "Enter any number "; cin >> i;
cout << "A-B=" << (a-b);
cout << "Now A = " << a << "and B = " << b;
cout << "Enter any number "; cin >> i;
cout << "A*B=" << (a*b);
cout << "Now A = " << a << "and B = " << b;
}
{
cout << "Enter any number "; cin >> i;
cout << "-A=" << (-a);
cout << "Now A = " << a;
cout << "-A=" << (-a);
cout << "Now A = " << a;
cout << "Enter any number "; cin >> i;
}
{
vector<bigint> cp = charpoly(a);
cout << "char. poly. of A has coefficients " << cp << endl;
cout << "det(A) = " << determinant(a) << endl;
}
{
aug = colcat(a,midmat(r));
cout << "Augmented matrix = " << aug << endl;
}

long rk, ny;
bigint denom;
{
int method;
cout << "Which echelon method? (0=standard,1=longlong,2=modular) ";
cin>>method;
ref = echelon(aug, pc, npc, rk, ny, denom, method);
cout << "Echelon matrix = " << ref;
cout << "pivotal columns: " << pc << endl;
cout << "nonpivotal columns: " << npc << endl;
cout << "Denom = " << denom << endl;
}

for (i=1,rk=0; (i<=r)&&(pc[i]<=r); i++,rk++) ;
ny = r-rk;
cout << "Rank = " << rk << endl;
cout << "Nullity = " << ny << endl;
if (rk<r) // non-invertible!
 {cout << "A is not invertible; rk = " << rk << endl;
 }
else
 {mat_m ainv = ref.slice(1,r,r+1,r+r);
  cout << "A has inverse ";
  if (!is_one(denom)) cout << "(1/" << denom << ")*";
  cout << ainv;
  cout << "Check: A.A^(-1) = I ?";
  if (a*ainv == denom*midmat(r)) cout << " True!";
  else cout << " False!";
  cout << endl;
 }
// Clear up:
 ref.init(); 
 pc.init(); npc.init(); 
 aug.init(); 
 a.init(); 
}
time(&stoptime);
cout << "cpu time = " << (stoptime-starttime) << " seconds\n";
}
