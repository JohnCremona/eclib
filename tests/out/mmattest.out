
Multiprecision matrix package test program.

Enter size of a square matrix A: Enter entries of A: A = 
[[1,2,3],
[4,5,6],
[7,8,9]]
Creating an array of 3 matrices
 A=
[[1,2,3],
[4,5,6],
[7,8,9]]
2A=
[[2,4,6],
[8,10,12],
[14,16,18]]
3A=
[[3,6,9],
[12,15,18],
[21,24,27]]
row(A,1) = [1,2,3]
row(A,2) = [4,5,6]
row(A,3) = [7,8,9]
A = 
[[1,2,3],
[4,5,6],
[7,8,9]]
col(A,1) = [1,4,7]
col(A,2) = [2,5,8]
col(A,3) = [3,6,9]
A = 
[[1,2,3],
[4,5,6],
[7,8,9]]
directsum(A,A) = 
[[1,2,3,0,0,0],
[4,5,6,0,0,0],
[7,8,9,0,0,0],
[0,0,0,1,2,3],
[0,0,0,4,5,6],
[0,0,0,7,8,9]]
Enter any number After shortening to a matrix of longs, A = 
[[1,2,3],
[4,5,6],
[7,8,9]]
B = A = 
[[1,2,3],
[4,5,6],
[7,8,9]]
Enter any number B==A?1
B!=A?0
after B+:=A, A = 
[[1,2,3],
[4,5,6],
[7,8,9]]
and B = 
[[2,4,6],
[8,10,12],
[14,16,18]]
Enter any number after B-:=A, A = 
[[1,2,3],
[4,5,6],
[7,8,9]]
and B = 
[[1,2,3],
[4,5,6],
[7,8,9]]
Enter any number after B*:=2, A = 
[[1,2,3],
[4,5,6],
[7,8,9]]
and B = 
[[2,4,6],
[8,10,12],
[14,16,18]]
Enter any number after B/:=2, A = 
[[1,2,3],
[4,5,6],
[7,8,9]]
and B = 
[[1,2,3],
[4,5,6],
[7,8,9]]
Enter any number A+B=
[[2,4,6],
[8,10,12],
[14,16,18]]
Now A = 
[[1,2,3],
[4,5,6],
[7,8,9]]
and B = 
[[1,2,3],
[4,5,6],
[7,8,9]]
Enter any number A-B=
[[0,0,0],
[0,0,0],
[0,0,0]]
Now A = 
[[1,2,3],
[4,5,6],
[7,8,9]]
and B = 
[[1,2,3],
[4,5,6],
[7,8,9]]
Enter any number A*B=
[[30,36,42],
[66,81,96],
[102,126,150]]
Now A = 
[[1,2,3],
[4,5,6],
[7,8,9]]
and B = 
[[1,2,3],
[4,5,6],
[7,8,9]]
Enter any number -A=
[[-1,-2,-3],
[-4,-5,-6],
[-7,-8,-9]]
Now A = 
[[1,2,3],
[4,5,6],
[7,8,9]]
-A=
[[-1,-2,-3],
[-4,-5,-6],
[-7,-8,-9]]
Now A = 
[[1,2,3],
[4,5,6],
[7,8,9]]
Enter any number char. poly. of A has coefficients [ 0 -18 -15 1 ]
det(A) = 0
Augmented matrix = 
[[1,2,3,1,0,0],
[4,5,6,0,1,0],
[7,8,9,0,0,1]]

Which echelon method? (0=standard,1=longlong,2=modular) Echelon matrix = 
[[3,0,-3,0,-8,5],
[0,3,6,0,7,-4],
[0,0,0,3,-6,3]]
pivotal columns: [1,2,4]
nonpivotal columns: [3,5,6]
Denom = 3
Rank = 2
Nullity = 1
A is not invertible; rk = 2
cpu time = 0 seconds
